lib/bugsnag/stacktrace.rb

```diff
@@ -10,105 +11,63 @@ class Stacktrace
     ##
     # Process a backtrace and the configuration into a parsed stacktrace.
     #
-    # rubocop:todo Metrics/CyclomaticComplexity
-    def initialize(backtrace, configuration)
-      @configuration = configuration
+    # @param backtrace [Array, nil] If nil, 'caller' will be used instead
+    # @param configuration [Configuration]
+    # @return [Array]
+    def self.process(backtrace, configuration) # rubocop:todo Metrics/CyclomaticComplexity
+      code_extractor = CodeExtractor.new(configuration)
 
       backtrace = caller if !backtrace || backtrace.empty?
 
-      @processed_backtrace = backtrace.map do |trace|
+      processed_backtrace = backtrace.map do |trace|
         # Parse the stacktrace line
         if trace.match(BACKTRACE_LINE_REGEX)
           file, line_str, method = [$1, $2, $3]
         elsif trace.match(JAVA_BACKTRACE_REGEX)
           method, file, line_str = [$1, $2, $3]
         end
 
-        next(nil) if file.nil?
+        next if file.nil?
 
         # Expand relative paths
         file = File.realpath(file) rescue file
 
         # Generate the stacktrace line hash
-        trace_hash = {}
-        trace_hash[:lineNumber] = line_str.to_i
+        trace_hash = { lineNumber: line_str.to_i }
 
-        if configuration.send_code
-          trace_hash[:code] = code(file, trace_hash[:lineNumber])
-        end
+        # Save a copy of the file path as we're about to modify it but need the
+        # raw version when extracting code (otherwise we can't open the file)
+        raw_file_path = file.dup
 
         # Clean up the file path in the stacktrace
-        if defined?(@configuration.project_root) && @configuration.project_root.to_s != ''
-          trace_hash[:inProject] = true if file.start_with?(@configuration.project_root.to_s)
-          file.sub!(/#{@configuration.project_root}\//, "")
-          trace_hash.delete(:inProject) if file.match(@configuration.vendor_path)
+        if defined?(configuration.project_root) && configuration.project_root.to_s != ''
+          trace_hash[:inProject] = true if file.start_with?(configuration.project_root.to_s)
+          file.sub!(/#{configuration.project_root}\//, "")
+          trace_hash.delete(:inProject) if file.match(configuration.vendor_path)
         end
 
-
         # Strip common gem path prefixes
         if defined?(Gem)
           file = Gem.path.inject(file) {|line, path| line.sub(/#{path}\//, "") }
         end
 
+        next if !file || file.empty?
```

Comment:
What case is this handling? The file path was equal to the project root / a gem path?

Reply:
Yeah exactly — [this test triggers it](https://github.com/bugsnag/bugsnag-ruby/blob/0fe026245b5e94939a3a407b856d8e1790dff56d/spec/stacktrace_spec.rb#L404-L425)

Reply:
Right but is that a thing that can happen? Backtrace lines that point to directories? If so, in the case that a line has a (non project root / gem path) directory, that should be something we think about in a few places.

Reply:
I don't think so, no. You could trigger it by setting the project root to a file rather than a directory too, which is a _little_ more realistic, but I don't think this is ever really triggered in practice

The only reason I kept it is because it [was here before](https://github.com/bugsnag/bugsnag-ruby/blob/59254960b973cbb8f635c29fe18526663b6c6a2c/lib/bugsnag/stacktrace.rb#L60-L64) and I can't say for certain that there's no reason for it. I went back through the history again and it was [implemented in this commit](https://github.com/bugsnag/bugsnag-ruby/commit/a479b6ed10ffdd3b808883037db92b27553aec6a) which also let users add their own filters for the stacktrace file name, so it would have been necessary then but seems like it's not anymore (unless the project root/a gem path point to a file)

Reply:
Hmmm ghosts in the code. I'd also suspect that its not needed anymore, especially since there are simpler mechanisms now for filtering out stacktrace lines if needed.

Reply:
<img src="https://media.giphy.com/media/etYZDbzfnrsvm/giphy.gif" height="64" alt="spooky">

👍 I'll remove it